File: quizzes3/09-problem.ts
export {};

const strRevWithLoop = (str: string): string => {
    let newString = '';

    for (let i = str.length - 1; i >= 0; i--) {
        newString += str[i];
    }

    return newString;
};

const strRevWithLoopEfficient = (str: string): string => {
    const result = new Array<string>();

    for (let i = 0, j = str.length - 1; j >= 0; i++, j--) {
        result[i] = str[j];
    }

    return result.join('');
};

const strRevWithChained = (str: string): string => {
    return str.split('').reverse().join('');
};

const strRevWithSpread = (str: string): string => {
    return [...str].reverse().join('');
};

console.log(JSON.stringify({ data: strRevWithLoop('hello') }, null, 2));
console.log(
    JSON.stringify({ data: strRevWithLoopEfficient('hello') }, null, 2),
);
console.log(JSON.stringify({ data: strRevWithChained('world') }, null, 2));
console.log(JSON.stringify({ data: strRevWithSpread('programmer') }, null, 2));

File: bins/binsV2/bin/bins1/09-problem.rb
# frozen_string_literal: true

require 'json'

def str_rev_with_loop(str)
  result = ''

  (str.length - 1).downto(0) do |i|
    result += str[i]
  end

  result
end

def str_rev_with_loop_efficient(str)
  chars = Array.new(str.length)
  j = str.length - 1

  str.length.times do |i|
    chars[i] = str[j]
    j -= 1
  end

  chars.join
end

def str_rev_with_chain(str)
  str.chars.reverse.join
end

def str_rev_with_spread(str)
  str.chars.reverse.join
end

puts JSON.pretty_generate({ data: str_rev_with_loop('hello') })
puts JSON.pretty_generate({ data: str_rev_with_loop_efficient('hello') })
puts JSON.pretty_generate({ data: str_rev_with_chain('world') })
puts JSON.pretty_generate({ data: str_rev_with_spread('programmer') })